pkg/provider/aws/secretsmanager/fake/fake.go

```diff
@@ -22,15 +23,35 @@ import (
 
 // Client implements the aws secretsmanager interface.
 type Client struct {
-	valFn func(*awssm.GetSecretValueInput) (*awssm.GetSecretValueOutput, error)
+	ExecutionCounter int
+	valFn            map[string]func(*awssm.GetSecretValueInput) (*awssm.GetSecretValueOutput, error)
 }
 
 func (sm *Client) GetSecretValue(in *awssm.GetSecretValueInput) (*awssm.GetSecretValueOutput, error) {
-	return sm.valFn(in)
+	sm.ExecutionCounter++
+	if entry, found := sm.valFn[sm.cacheKeyForInput(in)]; found {
+		return entry(in)
+	}
+	return nil, fmt.Errorf("test case not found")
+}
+
+func (sm *Client) cacheKeyForInput(in *awssm.GetSecretValueInput) string {
+	var secretID, versionID string
+	if in.SecretId != nil {
+		secretID = *in.SecretId
+	}
+	if in.VersionId != nil {
+		versionID = *in.VersionId
+	}
+	return fmt.Sprintf("%s#%s", secretID, versionID)
 }
 
 func (sm *Client) WithValue(in *awssm.GetSecretValueInput, val *awssm.GetSecretValueOutput, err error) {
-	sm.valFn = func(paramIn *awssm.GetSecretValueInput) (*awssm.GetSecretValueOutput, error) {
+	if sm.valFn == nil {
```

Comment:
this could go inside the init logic contained in tests, I put it here to avoid messing too much with original tests. However I do not have a strong opinion about this and can move the init logic elsewhere.

Reply:
I'd suggest adding a  `func NewClient() *Client` to the `fake` package that acts as a constructor for the Client type.

The `secretsmanager_test.go` can construct the client with `fakesm.NewClient()` instead of doing it directly with `&fakesm.Client{}`. This way we can keep the `WithValue` func clean and we have a clear way of constructing the Client type.

Reply:
fair point. Let me amend it.